pkg/engine/context/context.go

```diff
@@ -424,3 +427,41 @@ func (ctx *context) AddDeferredLoader(dl DeferredLoader) error {
 	ctx.deferred.Add(dl, len(ctx.jsonRawCheckpoints))
 	return nil
 }
+
+func (ctx *context) CreateDeepCopy() Interface {
+	if ctx == nil {
+		return nil
+	}
+
+	out := new(context)
+
+	out.jp = ctx.jp
+	out.operation = ctx.operation
+
+	if ctx.jsonRaw != nil {
+		out.jsonRaw = ctx.copyContext(ctx.jsonRaw)
+	}
+
+	if len(ctx.jsonRawCheckpoints) != 0 {
```

Comment:
do we need to copy checkpoints?

Seems like the only aspect of a context we need to deep copy is the JSON data.

Reply:
jsonCheckpoint map is used to create and reset checkpoints, if we share that object, I think that will cause another conflict where two go routines try to add/reset checkpoint at the same time and it will cause wrong behaviour